### PR TITLE
CameraXBasic updated to use 1.0.0-alpha08

### PR DESCRIPTION
**Pull request by syslogic**
This is the most recent version from the previous code review.

### CODE REVIEW

File: CameraXBasic/app/src/main/res/layout/fragment_camera.xml [hunk 1]
```diff
@@ -21,7 +21,7 @@
     android:layout_width="match_parent"
     android:layout_height="match_parent">
 
-    <TextureView
+    <androidx.camera.view.PreviewView
```
Comment (dharmendrainnovify) [inline CameraXBasic/app/src/main/res/layout/fragment_camera.xml]: @syslogic I agree that the PreviewView is coming from `androidx.camera:camera-view:1.0.0-alpha05` but that is not related to the `v1.0.0-alpha08` migration so can't we create separate PR for that?

Comment (syslogic) [reply CameraXBasic/app/src/main/res/layout/fragment_camera.xml]: In order to revert that, I'd have to add the `AutoFitPreviewBuilder` back in and it likely would break some more.

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt [hunk 2]
```diff
@@ -397,31 +416,30 @@ class CameraFragment : Fragment() {
...
+            bindCameraUseCases()
         }
 
-        // Listener for button used to view last photo
+        // Listener for button used to view the most recent photo
         controls.findViewById<ImageButton>(R.id.photo_view_button).setOnClickListener {
-            Navigation.findNavController(requireActivity(), R.id.fragment_container).navigate(
-                    CameraFragmentDirections.actionCameraToGallery(outputDirectory.absolutePath))
+            // Only navigate when the gallery has photos
+            if(outputDirectory.listFiles()?.size!! > 0) {
```
Comment (owahltinez) [inline CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: I would replace this with:
```
if (true == outputDirectory.listFiles()?.isNotEmpty())
```

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt [hunk 3]
```diff
@@ -397,31 +416,30 @@ class CameraFragment : Fragment() {
...
+                try {
+                    val dest = CameraFragmentDirections
+                            .actionCameraToGallery(outputDirectory.absolutePath)
+                    findNavController().navigate(dest)
```
Comment (owahltinez) [inline CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: Please change back to what it was for consistency:
```
Navigation.findNavController(requireActivity(), R.id.fragment_container).navigate(
        CameraFragmentDirections.actionCameraToGallery(outputDirectory.absolutePath))
```

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt [hunk 4]
```diff
@@ -397,31 +416,30 @@ class CameraFragment : Fragment() {
...
+                } catch(exc: IllegalArgumentException) {
```
Comment (owahltinez) [inline CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: I don't think that we need to catch the exception here, errors should be handled by the navigation library.

File: CameraXBasic/app/src/main/res/layout/fragment_camera.xml (hunk 1 above)
Comment (owahltinez) [reply CameraXBasic/app/src/main/res/layout/fragment_camera.xml]: `PreviewView` is the preferred `View` that should be used with `CameraX`. It simplifies user code a lot, see e.g. `AutofitTextureView` is no longer needed

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt [hunk 5]
```diff
@@ -132,7 +129,6 @@ class CameraFragment : Fragment() {
         override fun onDisplayChanged(displayId: Int) = view?.let { view ->
             if (displayId == this@CameraFragment.displayId) {
                 Log.d(TAG, "Rotation changed: ${view.display.rotation}")
-                preview?.setTargetRotation(view.display.rotation)
```
Comment (owahltinez) [inline CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: Because we are removing `AutoFitTextureView`, which handled the rotations, the preview is now broken after the device is rotated. It's not an easy fix, and eventually it should be solved by `PreviewView`, so I suggest removing the `android:configChanges` tag from the activity in the Manifest for now.

File: CameraXBasic/app/src/main/res/layout/fragment_camera.xml (hunk 1 above)
Comment (syslogic) [reply CameraXBasic/app/src/main/res/layout/fragment_camera.xml]: @owahltinez could this be the reason why `.setTargetRotation()` cannot be used?

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt [hunk 6]
```diff
@@ -277,60 +281,73 @@ class CameraFragment : Fragment() {
...
+                .also {
+                    it.setAnalyzer(mainExecutor, LuminosityAnalyzer {luma ->
                         // Values returned from our analyzer are passed to the attached listener
-                        // We log image analysis results here --
-                        // you should do something useful instead!
-                        val fps = (analyzer as LuminosityAnalyzer).framesPerSecond
-                        Log.d(TAG, "Average luminosity: $luma. " +
-                                "Frames per second: ${"%.01f".format(fps)}")
+                        // We log image analysis results here - you should do something useful instead!
+                        Log.d(TAG, "Average luminosity: $luma")
```
[the remark this answers is not in the document]
Comment (syslogic) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: Have closed the previous PR because of too many reverted commits, in order to provide a better overview and not to have all of them in the commit log. The reason why I removed the FPS is because it only returned value `infinite` and not any useful numeric value. If I'd revert this, I'd have to set the analyzer in two steps again, so that I could get a handle.

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt (hunk 4 above)
Comment (syslogic) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: This exception might not happen anymore, due to the check in line `431`.

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt (hunk 5 above)
Comment (syslogic) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: The `PreviewView` seems to lack an instance method `.setTargetRotation()` - only the builder has it.

Comment (syslogic) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: Have removed `android:configChanges`. When clicking the tiny rotate-icon on the system UI, it changes the layout orientation - just not automatically (which might be the intended behavior).

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt [hunk 7]
```diff
@@ -187,37 +183,38 @@ class CameraFragment : Fragment() {
...
             // If the folder selected is an external media directory, this is unnecessary
             // but otherwise other apps will not be able to access our images unless we
             // scan them using [MediaScannerConnection]
-            val mimeType = MimeTypeMap.getSingleton()
-                    .getMimeTypeFromExtension(photoFile.extension)
-            MediaScannerConnection.scanFile(
-                    context, arrayOf(photoFile.absolutePath), arrayOf(mimeType), null)
+            val mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(photoFile.extension)
+            // It prevents a force close, when pressing the photo-view button early on.
+            if(!requireActivity().isFinishing) {
```
Comment (owahltinez) [inline CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: I'm still unable to reproduce this issue. You mentioned that this gets triggered by pressing the shutter button, but that alone should not be finishing the activity. I would remove this check and, if you are able to provide steps to reliably reproduce this, we can discuss a fix.

Comment (syslogic) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: Not the shutter button, that photo-view / gallery button ...probably that "service connection lost" only may happen, when there are a few more images to scan, so that it would take more time to scan them. I noticed this when fixing the other other on `IllegalArgumentException`; my fault that I didn't take a note on how to crash it.

Comment (owahltinez) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: I also tried pressing the gallery button as quickly as possible upon startup, and I still can't reproduce. I suggest tracking this as a separate issue.

Comment (syslogic) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: It's `onImageSaved()`, so it's a) first the shutter button and b) then the photo-view button.

Comment (owahltinez) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: I still can't reproduce. I am seeing an error for the photo capture failing, because the camera closes before the capture completes: `androidx.camera.core.CameraClosedException: Camera is closed.`

So we can potentially disable the gallery button until capture finishes. Either way, I suggest tracking this as a separate issue.

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt (hunk 2 above)
Comment (syslogic) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: I could reproduce the `IllegalArgumentException` twice in a row (by wildly clicking buttons):

> IllegalArgumentException: navigation destination com.android.example.cameraxbasic:id/action_camera_to_gallery is unknown to this NavController

The other `if` condition might have been better.

File: CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt (hunk 7 above)
Comment (syslogic) [reply CameraXBasic/app/src/main/java/com/android/example/cameraxbasic/fragments/CameraFragment.kt]: Have removed it, if ever the error message will be "service connection lost", you'll know what it is.

**Comment by owahltinez** [comment]
Thank you @syslogic for this huge effort 🥇
